components/brave_wallet/browser/internal/hd_key_ed25519_slip23.cc

```diff
@@ -0,0 +1,237 @@
+  std::array<uint8_t, 32> public_key;
+  if (!ED25519_pubkey_from_scalar(public_key.data(), scalar.data())) {
+    return std::nullopt;
+  }
+  return public_key;
+}
+
+std::array<uint8_t, kSlip23ScalarSize> CalculateDerivedScalar(
+    base::span<const uint8_t, kSlip23ScalarSize> parent_scalar,
+    base::span<const uint8_t, kSlip23DerivationScalarSize> zl) {
+  std::array<uint8_t, kSlip23ScalarSize> out = {};
+
+  uint8_t carry = 0;
+  for (auto i = 0u; i < kSlip23DerivationScalarSize; i++) {
+    uint32_t r = parent_scalar[i] + (zl[i] << 3) + carry;
+    out[i] = r & 0xff;
+    carry = r >> 8;
+  }
+  for (auto i = kSlip23DerivationScalarSize; i < kSlip23ScalarSize; i++) {
+    uint32_t r = parent_scalar[i] + carry;
+    out[i] = r & 0xff;
+    carry = r >> 8;
+  }
+  return out;
+}
+
+std::array<uint8_t, kSlip23PrefixSize> CalculateDerivedPrefix(
+    base::span<const uint8_t, kSlip23PrefixSize> parent_prefix,
+    base::span<const uint8_t, kSlip23PrefixSize> zr) {
+  std::array<uint8_t, kSlip23PrefixSize> out = {};
+
+  uint8_t carry = 0;
+  for (auto i = 0u; i < kSlip23PrefixSize; i++) {
+    uint32_t r = parent_prefix[i] + zr[i] + carry;
+    out[i] = r;
+    carry = r >> 8;
+  }
+
+  return out;
+}
+
+std::array<uint8_t, kSlip23ChainCodeSize> CalculateDerivedChainCode(
+    base::span<const uint8_t, crypto::hash::kSha512Size> z_hmac) {
+  std::array<uint8_t, kSlip23ChainCodeSize> chain_code;
+  base::span(chain_code)
+      .copy_from(base::span(z_hmac).last<kSlip23ChainCodeSize>());
+  return chain_code;
+}
+
+}  // namespace
+
+HDKeyEd25519Slip23::HDKeyEd25519Slip23() = default;
+HDKeyEd25519Slip23::~HDKeyEd25519Slip23() = default;
+
+// Child key derivation constructor.
+HDKeyEd25519Slip23::HDKeyEd25519Slip23(
+    base::span<const uint8_t, kSlip23ScalarSize> scalar,
+    base::span<const uint8_t, kSlip23PrefixSize> prefix,
+    base::span<const uint8_t, kSlip23ChainCodeSize> chain_code,
+    base::span<const uint8_t, kEd25519PublicKeySize> public_key) {
+  base::span(scalar_).copy_from(scalar);
+  base::span(prefix_).copy_from(prefix);
+  base::span(chain_code_).copy_from(chain_code);
+  base::span(public_key_).copy_from(public_key);
+}
+
+std::unique_ptr<HDKeyEd25519Slip23> HDKeyEd25519Slip23::DeriveChild(
+    DerivationIndex index) {
+  auto raw_index_value = index.GetValue();
+  if (!raw_index_value) {
+    return nullptr;
+  }
+
+  std::array<uint8_t, crypto::hash::kSha512Size> z_hmac = {};
+  std::array<uint8_t, crypto::hash::kSha512Size> cc_hmac = {};
+
+  if (index.is_hardened()) {
+    std::array<uint8_t, 1 + 32 + 32 + 4> data;
+    auto span_writer = base::SpanWriter(base::span(data));
+    span_writer.Skip(1u);
+    span_writer.Write(scalar_);
+    span_writer.Write(prefix_);
+    span_writer.WriteU32LittleEndian(*raw_index_value);
+
+    data[0] = 0x00;
+    z_hmac = crypto::hmac::SignSha512(chain_code_, data);
+    data[0] = 0x01;
+    cc_hmac = crypto::hmac::SignSha512(chain_code_, data);
+  } else {
```

Comment:
Isn't this branch something we're not supposed to support according to SLIP10? That seems it was the way that we avoided any concerns for the paper that's broken and referenced in SLIP0023 (see [here](https://ieeexplore.ieee.org/document/7966967))

Reply:
SLIP10 does not support normal derivations but SLIP23 does